File: src/etch/examples/simpleecho/src/etch/examples/simpleecho/ClientMain.java
package etch.examples.simpleecho;

import java.lang.Thread;

import etch.bindings.java.transport.DefaultDeliveryService2;
import etch.bindings.java.transport.MailboxMessageSource;
import metreos.core.io.TcpConnection;

/**
 * Description of testRemoteFooServer.
 */
public class ClientMain 
{
    static StubSimpleEchoClient sfc;

    public static void main( String[] args ) throws Exception
    {
        ValueFactorySimpleEcho vf = new ValueFactorySimpleEcho();
        
        DefaultDeliveryService2<StubSimpleEchoClient> dds 
            = new DefaultDeliveryService2<StubSimpleEchoClient>()
        {
            @Override
            protected StubSimpleEchoClient newStub( MailboxMessageSource src )
            {
                return sfc;
            }

            @Override
            public String toString()
            {
                return "ClientMain";
            }
        };

        TcpConnection c = new TcpConnection( null, "", 4001, 0 );
        
        DefaultDeliveryService2 svc = 
            DefaultDeliveryService2.setUp(dds, c, vf );
        
        final SimpleEchoServer rfs = new RemoteSimpleEchoServer( svc );
        
        sfc = new StubSimpleEchoClient( new SimpleEchoClient()
        {
            public void onNewMessage(String message)
            {
                System.out.println(message); 
            }

            public void bogus() {}
        }, null, null );
        
        c.start();
        c.waitUp( 4000 );


        Thread t = new Thread() 
        {
            public void run() 
            {
                try 
                {
                    while (true)
                    {
                        rfs.write("Hi");
                        sleep(1000); 
                    }
                }
                catch(InterruptedException ex) {}
            }
        };

        System.out.println("Press enter to quit.");
        t.start();
        System.in.read();
        t.interrupt();
        t.join();
        System.out.println("Exiting.");

        svc.shutdownOutput();
        c.stop();
        c.waitDown(4000);
    }
}

File: src/etch/examples/simpleecho_cs/client/ClientMain.cs
using System;
using System.Collections.Generic;
using System.Text;

using Etch.Transport;
using Etch.Msg;

namespace etch.examples.simpleecho
{
    class ClientMain
    {
        public ClientDeliveryService cds = null;
        public TcpConnection c = null;
        public DefaultDeliveryService<StubSimpleEchoClient> svc = null;
        public ValueFactorySimpleEcho vf = null;
        public RemoteSimpleEchoServer rfs = null;
        public StubSimpleEchoClient sfc = null;

        public ClientMain()
        {
            vf = new ValueFactorySimpleEcho();
            c = new TcpConnection(null, "127.0.0.1", 4001, 0);
            sfc = new StubSimpleEchoClient(new SimpleEchoClientImpl(), null, null);
            cds = new ClientDeliveryService(this);
            svc = DefaultDeliveryService<StubSimpleEchoClient>.SetUp(cds, c, vf);
            rfs = new RemoteSimpleEchoServer(svc);
        }

        public class ClientDeliveryService : DefaultDeliveryService<StubSimpleEchoClient>
        {
            ClientMain cm = null;

            public ClientDeliveryService(ClientMain cm)
            {
                this.cm = cm;
            }

            protected override StubSimpleEchoClient NewStub(MailboxMessageSource src)
            {
                return cm.sfc;
            }

            public override string ToString()
            {
                return "ClientDeliveryService";
            }
        }

        static void Main(string[] args)
        {
            ClientMain cm = new ClientMain();

            cm.c.Start();
            cm.c.WaitUp(4000);

            cm.rfs.write("hi!!");
            System.Console.Read();

            cm.svc.ShutdownOutput();
            cm.c.Stop();
            cm.c.WaitDown(4000);
        }
    }
}
